File: erpnext/selling/page/pipe_balance_view/pipe_balance_view.py
import frappe
from erpnext.stock.report.pipe_balance.pipe_balance import get_data as get_report_data

@frappe.whitelist()
def get_dashboard_data(filters=None):
    if isinstance(filters, str):
        import json
        filters = json.loads(filters)
    
    data = get_report_data(filters)
    
    # Calculate summary stats
    total_weight = sum(d.get("weight", 0) for d in data)
    total_nos = sum(d.get("pieces", 0) for d in data)
    unique_packets = len(set(d.get("pkt_no") for d in data if d.get("pkt_no")))
    
    # Oldest received date
    dates = [d.get("date") for d in data if d.get("date")]
    oldest_date = min(dates) if dates else None
    
    return {
        "data": data,
        "stats": {
            "total_weight": total_weight,
            "total_nos": total_nos,
            "unique_packets": unique_packets,
            "oldest_date": oldest_date
        }
    }

@frappe.whitelist()
def export_to_excel(filters=None):
    if isinstance(filters, str):
        import json
        filters = json.loads(filters)
    
    data = get_report_data(filters)
    
    from collections import OrderedDict
    groups = OrderedDict()
    for row in data:
        pkt_no = row.get("pkt_no")
        if not pkt_no: continue
        if pkt_no not in groups:
            groups[pkt_no] = []
        groups[pkt_no].append(row)
    
    xlsx_data = []
    columns = [
        "PKT NO", "R DATE", "GRADE", "FINISH", "TYPE", "OD", "THIK", "LENGTH", "NOS", "WEIGHT", "PLOT", "L DAY"
    ]
    
    show_sold_stock = frappe.cint(filters.get("show_sold_stock")) if filters else 0
    for pkt_no, rows in groups.items():
        spec_groups = {}
        for row in rows:
            thk_str = str(row.get("thickness") or "").strip()
            try:
                val = float(thk_str)
                normalized_thk = str(int(val) if val.is_integer() else val)
            except ValueError:
                normalized_thk = thk_str.lower()

            spec_key = (
                row.get("grade") or "",
                row.get("finish") or "",
                row.get("type") or "",
                row.get("od") or "",
                normalized_thk,
                row.get("size") or ""
            )
            if spec_key not in spec_groups:
                spec_groups[spec_key] = []
            spec_groups[spec_key].append(row)

        balance_rows = []
        for spec_key, spec_rows in spec_groups.items():
            # Sort by date ascending to find oldest
            spec_rows.sort(key=lambda r: str(r.get("date") or ""))
            oldest_row = spec_rows[0]
            
            # Find latest row with positive weight for plot (warehouse)
            latest_active_row = oldest_row
            for r in reversed(spec_rows):
                if float(r.get("weight") or 0) > 0:
                    latest_active_row = r
                    break
            
            net_weight = sum(float(r.get("weight") or 0) for r in spec_rows)
            net_pieces = sum((-float(r.get("pieces") or 0) if float(r.get("weight") or 0) < 0 else float(r.get("pieces") or 0)) for r in spec_rows)
            
            finish = (oldest_row.get("finish") or "").strip().upper()
            is_2b_od = finish == "2B OD SIZE"
            is_sold = (net_weight <= 0.001) if is_2b_od else (net_weight <= 0.001 and net_pieces <= 0.001)

            if not show_sold_stock and is_sold:
                continue

            schedule = next((r.get("schedule") for r in spec_rows if r.get("schedule")), "")
            balance_rows.append((oldest_row, latest_active_row, schedule, net_pieces, net_weight))

        if not balance_rows:
            continue

        xlsx_data.append(columns)
        
        total_nos = 0
        total_weight = 0
        for oldest_row, latest_active_row, schedule, net_pieces, net_weight in balance_rows:
            thk = schedule or oldest_row.get("thickness") or ""
            if thk and "sch" not in str(thk).lower():
                try:
                    val = float(thk)
                    thk = f"{int(val) if val.is_integer() else val} mm"
                except ValueError:
                    if not str(thk).endswith("mm"):
                        thk = f"{thk} mm"

            xlsx_data.append([
                pkt_no,
                oldest_row.get("date"),
                oldest_row.get("grade"),
                oldest_row.get("finish"),
                oldest_row.get("type"),
                oldest_row.get("od"),
                thk,
                oldest_row.get("size"),
                net_pieces,
                net_weight,
                latest_active_row.get("plot"),
                oldest_row.get("l_days")
            ])
            total_nos += net_pieces
            total_weight += net_weight
            
        xlsx_data.append(["TOTAL", "", "", "", "", "", "", "", total_nos, total_weight, "", ""])
        xlsx_data.append([])
        xlsx_data.append([])

    from openpyxl import Workbook
    from openpyxl.styles import Font, PatternFill
    from io import BytesIO

    wb = Workbook()
    ws = wb.active
    ws.title = "Pipe Balance"
    
    bold_font = Font(bold=True)
    yellow_fill = PatternFill(start_color='FFFF00', end_color='FFFF00', fill_type='solid')

    for row_data in xlsx_data:
        ws.append(row_data)
        if not row_data:
            continue
        
        current_row = ws.max_row
        is_table_header = row_data[0] == "PKT NO"
        is_total_row = row_data[0] == "TOTAL"
        
        # Bold formatting
        if is_table_header or is_total_row:
            for cell in ws[current_row]:
                cell.font = bold_font
        
        # Yellow background for sold rows (Weight <= 0.001)
        # Skip headers and total rows
        elif not is_table_header and not is_total_row:
            if len(row_data) > 9 and isinstance(row_data[9], (int, float)) and row_data[9] <= 0.001:
                for cell in ws[current_row]:
                    cell.fill = yellow_fill

    xlsx_file = BytesIO()
    wb.save(xlsx_file)
    
    frappe.response['filename'] = f"Pipe_Balance.xlsx"
    frappe.response['filecontent'] = xlsx_file.getvalue()
    frappe.response['type'] = 'download'

File: erpnext/selling/page/pipe_balance_view/pipe_balance_view.js
frappe.pages['pipe-balance-view'].on_page_load = function (wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'Pipe Balance Dashboard',
		single_column: true
	});

	new PipeBalanceDashboard(page);
}

class PipeBalanceDashboard {
	constructor(page) {
		this.page = page;
		this.wrapper = $(page.body);
		this.filters = {};
		this.init();
	}

	init() {
		this.page.add_inner_button(__('Export'), () => {
			this.export_to_excel();
		});

		this.show_sold_stock = false;
		this.toggle_btn = this.page.add_inner_button(__('Show Sold Stock'), () => {
			this.show_sold_stock = !this.show_sold_stock;
			$(this.toggle_btn).toggleClass('btn-primary', this.show_sold_stock);
			this.refresh();
		});

		this.make_filters();
		this.make_layout();
		this.refresh();
	}

	make_filters() {
		this.page.add_field({
			fieldname: 'company',
			label: __('Company'),
			fieldtype: 'Link',
			options: 'Company',
			default: frappe.defaults.get_default('company'),
			change: () => this.refresh()
		});

		this.page.add_field({
			fieldname: 'warehouse',
			label: __('Warehouse'),
			fieldtype: 'Link',
			options: 'Warehouse',
			get_query: () => {
				return {
					filters: { 'is_group': 0 }
				}
			},
			change: () => this.refresh()
		});

		this.page.add_field({
			fieldname: 'to_date',
			label: __('To Date'),
			fieldtype: 'Date',
			default: frappe.datetime.get_today(),
			change: () => this.refresh()
		});
	}

	make_layout() {
		this.wrapper.html(`
			<div class="pipe-dashboard">
				<!-- Packet boxes will be rendered here -->
			</div>
		`);
	}

	refresh() {
		const filters = {
			company: this.page.fields_dict.company.get_value(),
			warehouse: this.page.fields_dict.warehouse.get_value(),
			to_date: this.page.fields_dict.to_date.get_value(),
			show_sold_stock: this.show_sold_stock ? 1 : 0
		};

		frappe.call({
			method: 'erpnext.selling.page.pipe_balance_view.pipe_balance_view.get_dashboard_data',
			args: { filters: filters },
			callback: (r) => {
				if (r.message) {
					this.render(r.message);
				}
			}
		});
	}

	export_to_excel() {
		const filters = {
			company: this.page.fields_dict.company.get_value(),
			warehouse: this.page.fields_dict.warehouse.get_value(),
			to_date: this.page.fields_dict.to_date.get_value(),
			show_sold_stock: this.show_sold_stock ? 1 : 0
		};

		open_url_post(
			'/api/method/erpnext.selling.page.pipe_balance_view.pipe_balance_view.export_to_excel',
			{ filters: JSON.stringify(filters) },
			true
		);
	}

	render(res) {
		const { data } = res;
		const body = $('#pipe-data-body');
		this.wrapper.find('.pipe-dashboard').empty();

		if (data.length === 0) {
			this.wrapper.find('.pipe-dashboard').append('<div style="text-align: center; padding: 20px;">No records found</div>');
			return;
		}

		// Group by Pkt No and preserve order
		const groups = new Map();
		data.forEach(row => {
			if (!row.pkt_no) return;
			if (!groups.has(row.pkt_no)) groups.set(row.pkt_no, []);
			groups.get(row.pkt_no).push(row);
		});

		groups.forEach((rows, pkt_no) => {
			const spec_groups = new Map();
			rows.forEach(row => {
				let normalized_thk = (row.thickness || '').trim();
				const parsed_thk = parseFloat(normalized_thk);
				if (!isNaN(parsed_thk)) {
					normalized_thk = String(parsed_thk);
				} else {
					normalized_thk = normalized_thk.toLowerCase();
				}

				const spec_key = `${row.grade || ''}_${row.finish || ''}_${row.type || ''}_${row.od || ''}_${normalized_thk}_${row.size || ''}`;
				if (!spec_groups.has(spec_key)) {
					spec_groups.set(spec_key, []);
				}
				spec_groups.get(spec_key).push(row);
			});

			let total_nos = 0;
			let total_weight = 0;
			const balance_rows = [];

			spec_groups.forEach((spec_rows, spec_key) => {
				// Sort by date ascending to find oldest
				spec_rows.sort((a, b) => {
					const date_a = a.date || '';
					const date_b = b.date || '';
					return date_a.localeCompare(date_b);
				});

				const oldest_row = spec_rows[0];

				// Find latest positive weight row for warehouse
				let latest_active_row = null;
				for (let i = spec_rows.length - 1; i >= 0; i--) {
					if (flt(spec_rows[i].weight) > 0) {
						latest_active_row = spec_rows[i];
						break;
					}
				}
				if (!latest_active_row) {
					latest_active_row = oldest_row;
				}

				const schedule_row = spec_rows.find(r => r.schedule && r.schedule.trim() !== '');
				const schedule = schedule_row ? schedule_row.schedule : '';

				let net_nos = 0;
				let net_weight = 0;
				spec_rows.forEach(row => {
					const weight = flt(row.weight);
					net_nos += weight < 0 ? -flt(row.pieces) : flt(row.pieces);
					net_weight += weight;
				});

				const finish_val = (oldest_row.finish || '').trim().toUpperCase();
				const is_2b_od = finish_val === '2B OD SIZE';
				const is_sold = is_2b_od ? (net_weight <= 0.001) : (net_weight <= 0.001 && net_nos <= 0.001);

				if (!this.show_sold_stock && is_sold) {
					return;
				}

				total_nos += net_nos;
				total_weight += net_weight;

				balance_rows.push({
					oldest_row: oldest_row,
					latest_active_row: latest_active_row,
					schedule: schedule,
					net_nos: net_nos,
					net_weight: net_weight
				});
			});

			if (balance_rows.length === 0) {
				return;
			}

			const rows_html = balance_rows.map(({ oldest_row, latest_active_row, schedule, net_nos, net_weight }) => {
				const row_class = net_weight <= 0.001 ? 'row-empty' : '';

				let date_str = '';
				if (oldest_row.date) {
					const date = frappe.datetime.str_to_obj(oldest_row.date);
					date_str = date.toLocaleDateString('en-US', { month: 'short', day: '2-digit' });
				}

				const l_days = oldest_row.date ? frappe.datetime.get_diff(frappe.datetime.get_today(), oldest_row.date) : 0;

				let thickness_display = schedule || oldest_row.thickness || '';
				if (thickness_display && !/sch/i.test(thickness_display)) {
					const val = parseFloat(thickness_display);
					if (!isNaN(val)) {
						thickness_display = `${val} mm`;
					} else if (!thickness_display.endsWith('mm')) {
						thickness_display = `${thickness_display} mm`;
					}
				}

				return `
					<tr class="${row_class}">
						<td style="font-weight: 700; color: var(--pipe-primary)">${pkt_no}</td>
						<td>${date_str}</td>
						<td>${oldest_row.grade || ''}</td>
						<td>${oldest_row.finish || ''}</td>
						<td>${oldest_row.type || ''}</td>
						<td>${oldest_row.od || ''}</td>
						<td>${thickness_display}</td>
						<td>${oldest_row.size || ''}</td>
						<td style="text-align: right">${net_nos}</td>
						<td style="text-align: right">${frappe.format(net_weight, { fieldtype: 'Float' }, { only_value: true })}</td>
						<td>${latest_active_row.plot || ''}</td>
						<td>${l_days}</td>
					</tr>
				`;
			}).join('');

			const box_html = `
				<div class="pkt-box" style="margin-bottom: 25px;">
					<div class="pipe-table-container">
						<table class="pipe-table">
							<thead>
								<tr>
									<th>PKT NO</th>
									<th style="width: 80px">R DATE</th>
									<th>GRADE</th>
									<th>FINISH</th>
									<th>TYPE</th>
									<th>OD</th>
									<th>THIK</th>
									<th>LENGTH</th>
									<th style="text-align: right">NOS</th>
									<th style="text-align: right">WEIGHT</th>
									<th>PLOT</th>
									<th>L DAY</th>
								</tr>
							</thead>
							<tbody>
								${rows_html}
								<tr style="background: #f8fafc; font-weight: 800">
									<td colspan="8">TOTAL</td>
									<td style="text-align: right">${total_nos}</td>
									<td style="text-align: right">${frappe.format(total_weight, { fieldtype: 'Float' }, { only_value: true })}</td>
									<td colspan="2"></td>
								</tr>
							</tbody>
						</table>
					</div>
				</div>
			`;
			this.wrapper.find('.pipe-dashboard').append(box_html);
		});
	}
}
